sprockets/mixins/amqp/__init__.py

```diff
@@ -120,68 +85,110 @@ def publish(self, exchange, routing_key, message, properties):
-        return pika.TornadoConnection(self._parameters,
-                                      self._on_connection_open,
-                                      self._on_connection_open_error,
-                                      custom_ioloop=ioloop.IOLoop.current())
+        self._connection = self._open_connection()
 
-    @property
-    def _connection_is_ready(self):
-        """Return True if the both the AMQP connection and channel are open
+    def _reconnect(self):
+        """Reconnect to RabbitMQ if the connection is closed and
+        not already connecting.
 
-        :rtype: bool
+        """
+        if self._connecting:
+            return
+        LOGGER.info('Reconnecting to %s', self._amqp_url)
+        self._io_loop.add_timeout(self._io_loop.time() + 0.1, self._connect)
+
+    def _open_connection(self):
+        """This method connects to RabbitMQ, returning the connection handle.
+        When the connection is established, the _on_connection_open method
+        will be invoked by pika. In the event the connection cannot be
+        opened, then the _on_connection_open_error is invoked.
+
+        :rtype: pika.TornadoConnection
 
         """
-        return self._connection_is_open and self._channel_is_open
+        LOGGER.info('Creating connection to %s', self._amqp_url)
+        return pika.TornadoConnection(
+            parameters=self._parameters,
+            on_open_callback=self._on_connection_open,
+            on_open_error_callback=self._on_connection_open_error,
+            custom_ioloop=ioloop.IOLoop.current())
 
     @property
-    def _connection_is_open(self):
-        """Returns ``True``if the connection to RabbitMQ is established and
-        ready to use.
+    def _parameters(self):
+        """Return a pika URLParameters object using the AMQP url
+        which identifies the Rabbit MQ server as a URL ready for
+        :class:`pika.connection.URLParameters`.
 
-        :rtype: bool
+        :rtype: pika.URLParameters
 
         """
-        return self._connection and self._connection.is_open
+        parameters = pika.URLParameters(self._amqp_url)
+        if self._connnection_attempts > 0:
+            parameters.connection_attempts = self._connnection_attempts
+        return parameters
 
     @property
-    def _channel_is_open(self):
+    def _is_ready(self):
         """Returns ``True``if the connection to RabbitMQ is established and
         ready to use.
 
         :rtype: bool
 
         """
-        return self.channel and self.channel.is_open
+        return self._channel and self._channel.is_open
 
-    @staticmethod
-    def _connection_timeout():
-        """Invoked when a connection to RabbitMQ has timed out and we
-        want to return an error to the client.
+    @gen.coroutine
+    def _connection_wait(self):
+        """Wait for a pending AMQP connection to complete"""
+        waiting_size = len(self._condition._waiters)
+        if waiting_size < self.MAX_WAITING_SIZE:
```

Comment:
Why are we limiting the number of waiters?  I'm particularly bothered by the usage of an internal detail (`_waiters` in this case) given that the Tornado maintainers tend to change internals without notice and don't have much sympathy for things that are broken in the wake of such changes.  I wouldn't add the complexity of managing the number of callers waiting on a condition without a pretty good reason.